### PR TITLE
buildsystem: printf conversion to function

### PR DESCRIPTION
**Pull request by antonlacon**
WIP demonstration of adding a printf helper function to config/functions to output build messages. Currently two functions, but easily combined to one (simply insert build_msg_indent's content into the if test where it's called). Since not every status message causes indent to shift further, this is controlled through the 4th parameter being "indent" or not (and erroring out on other content in $4)

get_archive conversion provided as sample. Tested by fetching a variety of packages.

### CODE REVIEW

File: scripts/get_archive [hunk 1]
```diff
@@ -37,7 +36,7 @@ while [ ${NBWGET} -gt 0 ]; do
 
       [ -z "${PKG_SHA256}" -o "${PKG_SHA256}" = "${CALC_SHA256}" ] && break 2
 
-      printf "%${BUILD_INDENT}c $(print_color CLR_WARNING "WARNING") Incorrect checksum calculated on downloaded file: got ${CALC_SHA256} wanted ${PKG_SHA256}\n\n" ' '>&$SILENT_OUT
+      build_msg "CLR_WARNING" "WARNING" "Incorrect checksum calculated on download file: got ${CALC_SHA256} wanted ${PKG_SHA256}"
```
Comment (MilhouseVH) [inline scripts/get_archive]: This is losing one `\n` - intentional? And same below.

**Comment by MilhouseVH** [comment]
Looks good, thanks - just a few minors.

File: scripts/get_archive (hunk 1 above)
Comment (antonlacon) [reply scripts/get_archive]: Intentional for both. I don't see value in the extra blank line. Easily readded as part of $3 if desired.

Comment (MilhouseVH) [reply scripts/get_archive]: Ok fine, let's see how it goes without the extra `\n` (sometimes I prefer a little bit of extra space for clarity).

**Comment by MilhouseVH** [comment]
Thanks for the changes. Now that we're centralising this code, it might be worth applying a couple of extra cleanups. Given the following:
```
build_msg() {
  printf "%${BUILD_INDENT}c $(print_color "${1}" "${2}")      ${3}\n" ' '>&${SILENT_OUT}

  # pad left space to create "indent" effect
  if [ "${4}" = "indent" ]; then
    export BUILD_INDENT=$((${BUILD_INDENT:-1}+${BUILD_INDENT_SIZE}))
...
```
1) When `BUILD_INDENT` is not defined, the indentation argument becomes `%c` which outputs zero spaces, therefore the default value for `BUILD_INDENT` should be 0
2) Given (1) above, I'm not sure why we default `BUILD_INDENT` to 1 with `${BUILD_INDENT:-1}` - should we use 0 instead? This would have a minor (but IMHO correct) cosmetic effect on indentation - instead of indentation at columns 5, 9, 13 etc. it will be at columns 4, 8, 12 etc.
3) I'm also not sure why we include a space between the first two output values `%${BUILD_INDENT}c $(print_color "${1}" "${2}")` - maybe `%${BUILD_INDENT}c$(print_color "${1}" "${2}")` would be better, and tighten up the output?

The following implements the changes mentioned above:
```
build_msg() {
  printf "%${BUILD_INDENT:-0}c$(print_color "${1}" "${2}")      ${3}\n" ' '>&${SILENT_OUT}

  # pad left space to create "indent" effect
  if [ "${4}" = "indent" ]; then
    export BUILD_INDENT=$((${BUILD_INDENT:-0}+${BUILD_INDENT_SIZE}))
...
```

---

There is also one example that doesn't call `print_color()`, in [scripts/clean](https://github.com/LibreELEC/LibreELEC.tv/blob/843051d0f78812ecca0d9593577dcbf671c96e98/scripts/clean#L27).

This could become:
```
-      printf "%${BUILD_INDENT}c * Removing $i ...\n" ' '>&$SILENT_OUT
+      build_msg "" "" "* Removing $i ..."
```
and then we will need to add checks for args 1 and 2 being blank - probably best to do this in `print_color()`:
```
diff --git a/config/functions b/config/functions
index 9588806..f5b772c 100644
--- a/config/functions
+++ b/config/functions
@@ -38,6 +38,8 @@ listremoveitem() {
 print_color() {
   local clr_name="$1" clr_text="$2" clr_actual

+  [ -z "${clr_name}" ] && return 0
+
   if [ "$DISABLE_COLORS" = "yes" ]; then
     [ $# -eq 2 ] && echo -en "${clr_text}"
     return 0
```

**Comment by antonlacon** [comment]
Thoughts on replacing the `export $BUILD_INDENT...` stuff with `$4` determining desired "tab" column? Looks like:

```
build_msg() {
  local left_pad=$((${4:-0}*${BUILD_INDENT_SIZE}))
  printf "%${left_pad}c $(print_color "${1}" "${2}")      ${3}\n" ' '>&${SILENT_OUT}
}
```

Normal message:
`build_msg "CLR_GET" "GET" "${1} (archive)"`

Indented message:
`build_msg "CLR_INFO" "INFO" "Calculated checksum: ${CALC_SHA256}" "1"`

**Comment by MilhouseVH** [comment]
The way it works at the moment is that the amount of indentation used will increment progressively, relative to the current level of indent, whereas your proposal would be converting the amount of indentation to an absolute value.

Absolute levels of indentation may be more difficult to work with as the indentation required by a script can vary depending on the context in which it is used/executed - for example when `scripts/clean` is called via `scripts/unpack` it will indent relative to the `unpack` messages (which itself will be relative to `scripts/build` etc.), but `scripts/clean` can also be executed directly from the command line where the level of indentation required is zero.

So on the whole I think I prefer the current implementation, as the script doesn't need to be concerned about the amount of indentation and it either wants additional indentation, or it doesn't.

**Comment by antonlacon** [comment]
Most of `$SCRIPTS` is converted. autoreconf and clean were done with general cleanups because they're short. get_file and get_git are untested as my grep skills couldn't locate a package that would use them in the tree. get_git also had an if test improvement or two, and create_addon fixes the device and arch test error messages being reversed.

I'll look at the remaining scripts in a day or two.

**Comment by antonlacon** [comment]
I think everything is converted now. Test built a rpi2 image and created binary addons and nothing threw an error to kill the builds. Caveats re get_file and get_git remain being untested remain.

File: scripts/build [hunk 2]
```diff
@@ -209,11 +209,10 @@ for p in $_pkg_depends; do
 done
 
 if [ "${BUILD_WITH_DEBUG}" = "yes" ]; then
-  printf "%${BUILD_INDENT}c $(print_color CLR_BUILD "BUILD")    $PACKAGE_NAME $(print_color CLR_TARGET "($TARGET)") [DEBUG]\n" ' '>&$SILENT_OUT
+  build_msg "CLR_BUILD" "BUILD" "${PACKAGE_NAME} $(print_color CLR_TARGET "(${TARGET})" [DEBUG])" "indent"
```
Comment (MilhouseVH) [inline scripts/build]: ` [DEBUG]` is not a parameter to `print_color()` - this line should be:
```suggestion
  build_msg "CLR_BUILD" "BUILD" "${PACKAGE_NAME} $(print_color CLR_TARGET "(${TARGET})") [DEBUG]" "indent"
```

File: scripts/build [hunk 3]
```diff
@@ -209,11 +209,10 @@ for p in $_pkg_depends; do
...
 else
-  printf "%${BUILD_INDENT}c $(print_color CLR_BUILD "BUILD")    $PACKAGE_NAME $(print_color CLR_TARGET "($TARGET)")\n" ' '>&$SILENT_OUT
+  build_msg "CLR_BUILD" "BUILD" "${PACKAGE_NAME} $(print_color CLR_TARGET "(${TARGET})")" "indent"
```
Comment (MilhouseVH) [inline scripts/build]: `"CLR_TARGET"` (just to be consistent)

File: scripts/create_addon [hunk 4]
```diff
@@ -113,12 +113,12 @@ pack_addon() {
       if [ "$ADDON_OVERWRITE" = "yes" ]; then
         rm $ADDON_INSTALL_DIR/$PKG_ADDON_ID-$ADDONVER.zip
       else
-        printf  "%${BUILD_INDENT}c $(print_color CLR_WARNING "*** WARNING: $PKG_ADDON_ID-$ADDONVER.zip already exists. not overwriting it ***")\n" ' '>&$SILENT_OUT
+        build_msg "CLR_WARNING" "" "*** WARNING: ${PKG_ADDON_ID}-${ADDONVER}.zip already exists. Not overwrittng it. ***"
```
Comment (MilhouseVH) [inline scripts/create_addon]:
```suggestion
        build_msg "CLR_WARNING" "*** WARNING: ${PKG_ADDON_ID}-${ADDONVER}.zip already exists. Not overwrittng it. ***" ""
```
as we want the message text to be output using the "warning" colour (ie. red).


We may want to check arg3 to `build_msg()` and use a slightly different printf when it is blank. For example:
```
-  printf "%${BUILD_INDENT}c$(print_color "${1}" "${2}")      ${3}\n" ' '>&${SILENT_OUT} 
+  if [ -n "${3}" ]; then
+    printf "%${BUILD_INDENT}c$(print_color "${1}" "${2}")      ${3}\n" ' '>&${SILENT_OUT}
+  else
+    printf "%${BUILD_INDENT}c$(print_color "${1}" "${2}")\n" ' '>&${SILENT_OUT}
+  fi
```

File: scripts/create_addon [hunk 5]
```diff
@@ -113,12 +113,12 @@ pack_addon() {
...
         return 0
       fi
     fi
     cd $ADDON_BUILD
-    printf  "%${BUILD_INDENT}c $(print_color CLR_INFO "*** compressing Addon $PKG_ADDON_ID ... ***")\n" ' '>&$SILENT_OUT
+    build_msg "CLR_INFO" "" "*** compressing addon ${PKG_ADDON_ID} ... ***"
```
Comment (MilhouseVH) [inline scripts/create_addon]: Same as the CLR_WARNING above - this needs to be:
```suggestion
    build_msg "CLR_INFO" "*** compressing addon ${PKG_ADDON_ID} ... ***" ""
```

**Comment by MilhouseVH** [comment]
Sorry for all the spam - I had intended for this all to be a single "review" but unintentionally began with single comments.

Just a few minors and nits, other than that it all looks good - thanks.

**Comment by antonlacon** [comment]
Minors addressed and new image built. No issues encountered.

**Comment by MilhouseVH** [comment]
I found an existing error in `scripts/build` that we should fix:

https://github.com/LibreELEC/LibreELEC.tv/blob/master/scripts/build#L268

there is no such color as `bold-red` - it's `boldred`.

I would suggest this should become:
```
  die "$(print_color "CLR_ERROR" "ERROR:") unknown toolchain $PKG_TOOLCHAIN"
```
using this [commit](https://github.com/LibreELEC/LibreELEC.tv/commit/f839437fa253d9ced4a835e67c661a62d77ce5c2)

**Comment by antonlacon** [comment]
Merged.

**Pull request merged**